smop/libscripts/octave-4.0.2/doc/interpreter/mkoctfile.1
.\" Copyright (C) 2000-2015 Dirk Eddelbuettel
.\"
.\" This file is part of Octave.
.\"
.\" Octave is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 3 of the License, or (at
.\" your option) any later version.
.\"
.\" Octave is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with Octave; see the file COPYING.  If not, see
.\" <http://www.gnu.org/licenses/>.
.\"
.\" This page was contributed by Dirk Eddelbuettel <edd@debian.org>
.\" This page was completely re-written by Rik <octave@nomad.inbox5.com>
.\" --------------------------------------------------------------------
.de Vb \" (V)erbatim (b)egin.  Use fixed width font and no justification
.ft CW
.nf
..
.de Ve \" (V)erbatim (e)nd.  Return to regular font and justification
.ft R
.fi
..
.\" --------------------------------------------------------------------
.TH MKOCTFILE 1 "3 January 2014" "GNU Octave"
.SH NAME
mkoctfile \- Compile dynamic-load modules for GNU Octave
.SH SYNOPSIS
\fBmkoctfile\fP [\fIOPTION\fP]... file...
.SH DESCRIPTION
\fBmkoctfile\fP is used to compile C, C++, or Fortran source code in
to a dynamically loadable \fI.oct\fP file for
.BR octave (1).
.SH OPTIONS
.TP 8
\fB\-h\fP, \fB\-?\fP, \fB\-\-help\fP
Print help message.
.TP
.B \-I\fIdir\fP
Add include directory \fIdir\fP to compile commands.
.TP
.B \-idirafter\fIdir\fP
Add include directory to second include search path after '\fB\-I\fP'.
.TP
.B \-D\fIdef\fP
Add definition \fIdef\fP to compiler command.
.TP
.B \-l\fIlib\fP
Add library \fIlib\fP to link command.
.TP
.B \-L\fIdir\fP
Add library directory \fIdir\fP to link command.
.TP
.B \-R\fIdir\fP
Pass \fB\-R\fP\fIdir\fP to link command.
.TP
.B \-g
Enable debugging option for all compilers.
.TP
.B \-pthread
Add \fB\-pthread\fP to link command.
.TP
.B \-W...
Pass flags to the compiler such as \fB\-Wa,\fP\fIOPTION\fP.
.TP
.B \-Wl,...
Pass flags to the linker such as \fB\-Wl,-rpath=...\fP
.TP
.B \-M\fR,\fB \-\-depend
Generate dependency files (.d) for C and C++ source files.
.TP
.B \-c
Compile but do not link.
.TP
.B \-\-link-stand-alone
Link a stand-alone executable file.
.TP
.B \-s\fR,\fB --strip
Strip the output file.
.TP
.B \-\-mex
Create a MEX file.  Set the default output extension to \fB.mex\fP.
.TP
.B \-o \fIfile\fP\fR,\fB \-\-output \fIfile\fP
Output file name.  Default extension is \fB.oct\fP (or \fB.mex\fP if \-\-mex is
specified) unless linking a stand-alone executable.
.TP
.B \-p \fIVAR\fP\fR,\fB \-\-print \fIVAR\fP
Print configuration variable \fIVAR\fP.  Recognized variables are:
.RS
.Vb
    ALL_CFLAGS                FFTW3F_LDFLAGS
    ALL_CXXFLAGS              FFTW3F_LIBS
    ALL_FFLAGS                FLIBS
    ALL_LDFLAGS               FPICFLAG
    AR                        INCFLAGS
    BLAS_LIBS                 LAPACK_LIBS
    CC                        LDFLAGS
    CFLAGS                    LD_CXX
    CPICFLAG                  LD_STATIC_FLAG
    CPPFLAGS                  LFLAGS
    CXX                       LIBOCTAVE
    CXXFLAGS                  LIBOCTINTERP
    CXXPICFLAG                LIBS
    DEPEND_EXTRA_SED_PATTERN  OCTAVE_LIBS
    DEPEND_FLAGS              OCTAVE_LINK_DEPS
    DL_LD                     OCT_LINK_DEPS
    DL_LDFLAGS                RANLIB
    EXEEXT                    RDYNAMIC_FLAG
    F77                       READLINE_LIBS
    F77_INTEGER_8_FLAG        SED
    FFLAGS                    XTRA_CFLAGS
    FFTW3_LDFLAGS             XTRA_CXXFLAGS
    FFTW3_LIBS
.Ve
.RE
.TP
.B \-v\fR,\fB --verbose
Echo commands as they are executed.
.TP
.B file
Compile or link file.  Recognized file types are
.RS
.Vb
   .c    C source
   .cc   C++ source
   .C    C++ source
   .cpp  C++ source
   .f    Fortran source (fixed form)
   .F    Fortran source (fixed form)
   .f90  Fortran source (free form)
   .F90  Fortran source (free form)
   .o    object file
   .a    library file
.Ve
.RE
.SH AUTHOR
John W. Eaton <jwe@octave.org>

This manual page was contributed by Dirk Eddelbuettel
<edd@debian.org> for the Debian GNU/Linux distribution but
may be used by others.
.SH "SEE ALSO"
octave (1).


NEWS.rst
=============
Release notes
=============

    Version 0.33 is out

    New option -c --code 
        Quickly check how a particular statement
        is translated. Option -c implies -o- .

    Renamed option -D to -d (--debug).

    New regression count:

        $ cd libscripts && make regresss

    Currently 14 regressions

        
November 18, 2016
    Version 0.31 is out

    Changed naming of output files.
        If -o not specified, compiling foo.m
	now results in foo.py

    Using octave script library for testing
        $ wget ftp://ftp.gnu.org/ftp/octave/octave-4.0.2.tar.gz
	$ smop -a octave-4.0.2.tar.gz -g '*/scripts/*.m'
	$ ls -1 *.py | wc
	$ python -m py_compile *.py
	$ ls -1 *.pyc | wc

    New option -a --archive allows to get data from
        tar file, without extracting the contents
	to a temporary place.

    New option -g --glob-pattern allows to limit
        the amount of input data

    Removed option --ignore-errors.
        All errors are ignored now.

    
August 27, 2016
    Version 0.29 is out

    Test suite for regression testing is ready,
        with about 1000 m-scripts.  Scripts are translated
        to python, then the resulting py-files are loaded,
        but not yet run.
    
August 11, 2016
    Version 0.28 is out. 

    Line numbering information is included in the output.
        Enabled by default.  Disabled with ``--no-numbers``.

    Block comments are preserved now.
        Lines, containing anything but a comment, are
        preserved. Enabled by default.
        Disabled with ``--no-comments``.

    New command-line options
        ``--no-resolve,--no-backend`` useful mostly for
        debugging. New shortcuts -R -B -T -C -N
        
    Special ``%!`` comments are partially supported now.
        They are mostly useful in testing, and
        are used by Octave library and test suite. Disabled
        by default. Enabled with ``--testing-mode``.
        

June 19,2016
   After a year-long vacation I am back to active development.
   My first goal is adopting the Octave runtime and test suite.

October 23, 2014
   Downloaded ``mybench`` -- a collection of 20 or so
   micro-benchmarks originally meant to compare matlab and
   octave performance.  After succesfully running the first nine,
   the geometric mean of the speedup is 0.36,  which is cool.

    ==   ========   ======    ===========    =======
    //   name       octave    smop           speedup
    ==   ========   ======    ===========    =======
    1    rand       2.58      0.36           0.14
    2    randn      2.26      1.04           0.46
    3    primes     0.35      0.17           0.49
    4    fft2       2.75      1.13           0.41
    5    square     4.24      0              
    6    inv        4.38      2.26           0.53
    7    eig        17.95     9.09           0.51
    8    qr         3.06      1.83           0.60
    9    shur       5.98      2.31           0.39
    10   roots      8.31      2.02           0.24
    ==   ========   ======    ===========    =======

October 15, 2014
   Version 0.26.3 is available for beta testing.
   Next version 0.27 is planned to compile octave
   ``scripts`` library, which contains over 120 KLOC in
   almost 1,000 matlab files. There  are 13 compilation
   errors with smop 0.26.3 .



smop/libscripts/octave-4.0.2/scripts/miscellaneous/mkoctfile.m
## Copyright (C) 2006-2015 Keith Goodman
##
## This file is part of Octave.
##
## Octave is free software; you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 3 of the License, or (at
## your option) any later version.
##
## Octave is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with Octave; see the file COPYING.  If not, see
## <http://www.gnu.org/licenses/>.

## -*- texinfo -*-
## @deftypefn  {Command} {} mkoctfile [-options] file @dots{}
## @deftypefnx {Function File} {[@var{output}, @var{status}] =} mkoctfile (@dots{})
##
## The @code{mkoctfile} function compiles source code written in C, C++, or
## Fortran.  Depending on the options used with @code{mkoctfile}, the
## compiled code can be called within Octave or can be used as a stand-alone
## application.
##
## @code{mkoctfile} can be called from the shell prompt or from the Octave
## prompt.  Calling it from the Octave prompt simply delegates the call to
## the shell prompt.  The output is stored in the @var{output} variable and
## the exit status in the @var{status} variable.
##
## @code{mkoctfile} accepts the following options, all of which are optional
## except for the file name of the code you wish to compile:
##
## @table @samp
## @item -I DIR
## Add the include directory DIR to compile commands.
##
## @item -D DEF
## Add the definition DEF to the compiler call.
##
## @item -l LIB
## Add the library LIB to the link command.
##
## @item -L DIR
## Add the library directory DIR to the link command.
##
## @item  -M
## @itemx --depend
## Generate dependency files (.d) for C and C++ source files.
##
## @item -R DIR
## Add the run-time path to the link command.
##
## @item @nospell{-Wl,@dots{}}
## Pass flags though the linker like @nospell{"-Wl,-rpath=@dots{}"}.
## The quotes are needed since commas are interpreted as command
## separators.
##
## @item -W@dots{}
## Pass flags though the compiler like @nospell{"-Wa,OPTION"}.
##
## @item -c
## Compile but do not link.
##
## @item -g
## Enable debugging options for compilers.
##
## @item  -o FILE
## @itemx --output FILE
## Output file name.  Default extension is .oct (or .mex if @samp{--mex} is
## specified) unless linking a stand-alone executable.
##
## @item  -p VAR
## @itemx --print VAR
## Print the configuration variable VAR@.  Recognized variables are:
##
## @example
##    ALL_CFLAGS                  INCFLAGS
##    ALL_CXXFLAGS                INCLUDEDIR
##    ALL_FFLAGS                  LAPACK_LIBS
##    ALL_LDFLAGS                 LD_CXX
##    AR                          LDFLAGS
##    BLAS_LIBS                   LD_STATIC_FLAG
##    CC                          LFLAGS
##    CFLAGS                      LIBDIR
##    CPICFLAG                    LIBOCTAVE
##    CPPFLAGS                    LIBOCTINTERP
##    CXX                         LIBS
##    CXXFLAGS                    OCTAVE_HOME
##    CXXPICFLAG                  OCTAVE_LIBS
##    DEPEND_EXTRA_SED_PATTERN    OCTAVE_LINK_DEPS
##    DEPEND_FLAGS                OCTAVE_LINK_OPTS
##    DL_LD                       OCTAVE_PREFIX
##    DL_LDFLAGS                  OCTINCLUDEDIR
##    F77                         OCTLIBDIR
##    F77_INTEGER8_FLAG           OCT_LINK_DEPS
##    FFLAGS                      OCT_LINK_OPTS
##    FFTW3F_LDFLAGS              RANLIB
##    FFTW3F_LIBS                 RDYNAMIC_FLAG
##    FFTW3_LDFLAGS               READLINE_LIBS
##    FFTW3_LIBS                  SED
##    FFTW_LIBS                   SPECIAL_MATH_LIB
##    FLIBS                       XTRA_CFLAGS
##    FPICFLAG                    XTRA_CXXFLAGS
## @end example
##
## @item --link-stand-alone
## Link a stand-alone executable file.
##
## @item --mex
## Assume we are creating a MEX file.  Set the default output extension to
## ".mex".
##
## @item  -s
## @itemx --strip
## Strip the output file.
##
## @item  -v
## @itemx --verbose
## Echo commands as they are executed.
##
## @item file
## The file to compile or link.  Recognized file types are
##
## @example
## @group
##    .c    C source
##    .cc   C++ source
##    .C    C++ source
##    .cpp  C++ source
##    .f    Fortran source (fixed form)
##    .F    Fortran source (fixed form)
##    .f90  Fortran source (free form)
##    .F90  Fortran source (free form)
##    .o    object file
##    .a    library file
## @end group
## @end example
##
## @end table
## @end deftypefn

function [output, status] = mkoctfile (varargin)

  bindir = octave_config_info ("bindir");
  ext = octave_config_info ("EXEEXT");

  shell_script = fullfile (bindir,
                           sprintf ("mkoctfile-%s%s", OCTAVE_VERSION, ext));

  if (! exist (shell_script, "file"))
    __gripe_missing_component__ ("mkoctfile", "mkoctfile");
  endif

  cmd = ['"' shell_script '"'];
  for i = 1:nargin
    cmd = [cmd ' "' varargin{i} '"'];
  endfor

  [sys, out] = system (cmd);

  if (nargout > 0)
    [output, status] = deal (out, sys);
  else
    printf ("%s", out);
  endif

  if (sys != 0)
    warning ("mkoctfile exited with failure status");
  endif

endfunction

